### PR TITLE
utils: implement rt module

### LINKED ISSUE #2418: Create RT API
Utility for RT to be used across all modules (editor and holdingpen)

## Needed actions

- get all tickets by `RecordID`
- close a ticket `ticket_id`
- get RT users
- get Queue names
- create a ticket
- ~~update a ticket by `ticket_id`~~
- reply a ticket by `ticked_id`

### PR DESCRIPTION
**Pull request by harunurhan**
* Implements Rt module (fixes #2418)

### CODE REVIEW

File: inspirehep/utils/tickets.py [hunk 1]
```diff
@@ -24,26 +24,222 @@
...
+    """Mapper function
+
+    Takes a string like ``'17: CoolUser'`` and
+    returns ``{'id': '17', 'name': 'CoolUser'}``
+    """
+    parts = raw_item.split(": ")
+    return dict(id=parts[0], name=parts[1])
+
+
+def _strip_lines(multiline_string):
```
Comment (jmartinm) [inline inspirehep/utils/tickets.py]: A test for this would be useful

File: inspirehep/utils/tickets.py [hunk 2]
```diff
@@ -24,26 +24,222 @@
...
+    return "\n ".join(
+        [line.strip() for line in multiline_string.strip().split("\n")]
+    )
+
+
+def get_tickets_by_recid(recid):
+    """Returns all tickets that is associated with the given recid
```
Comment (jmartinm) [inline inspirehep/utils/tickets.py]: `that are`

File: inspirehep/utils/ext.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        url = app.config.get("CFG_BIBCATALOG_SYSTEM_RT_URL", "")
+        login = app.config.get("CFG_BIBCATALOG_SYSTEM_RT_DEFAULT_USER", "")
+        password = app.config.get(
+            "CFG_BIBCATALOG_SYSTEM_RT_DEFAULT_PWD", "")
+        if url:
+            tracker = Rt(
+                url=url,
+                default_login=login,
+                default_password=password,
+                verify_cert=False,
```
Comment (harunurhan) [inline inspirehep/utils/ext.py]: yes, but we also don't need it,  so I was gonna remove when I am done with it and don't need testing.

File: inspirehep/utils/ext.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            )
+            tracker.login()
```
Comment (harunurhan) [inline inspirehep/utils/ext.py]: `login` saves token to `request.session` cookies. It should be ok during application life time, but I will try to check how likely this token is expired in Rt.

File: inspirehep/utils/tickets.py [hunk 5]
```diff
@@ -24,26 +24,222 @@
...
+from os.path import join
+
+from flask import render_template
+
+from rt import ALL_QUEUES, ConnectionError
+
+from .proxies import rt_instance
 
 
 def get_instance():
```
Comment (harunurhan) [inline inspirehep/utils/tickets.py]: this is not the break the existing code, after refactoring the code base, I will remove.

File: inspirehep/utils/tickets.py [hunk 6]
```diff
@@ -24,26 +24,222 @@
...
+def get_users():
+    """Returns list of all users as {id, name} dict
+
+    :rtype: dict - with ``name (string)``, ``id (integer)`` properties
+    """
+    return _get_all_of("user")
+
+
+def _get_all_of(query_type):
+    """Utility function to share the code for performing custom search queries
```
Comment (harunurhan) [inline inspirehep/utils/tickets.py]: yes that's true!, need to modify the docstring a bit.

File: inspirehep/utils/tickets.py [hunk 7]
```diff
@@ -24,26 +24,222 @@
...
     return isinstance(exception, ConnectionError)
+
+
+def create_ticket(queue, requestors, subject, body, recid=None, **kwargs):
+    """Creates new RT ticket and returns new ticket id.
+
+    :param queue: queue name for the ticket
+    :type queue: string
+
+    :param requestors: username to set to requestors field of the ticket
```
Comment (harunurhan) [inline inspirehep/utils/tickets.py]: yep, that's an RT thing, and it's a bit weird. Actually their rest documentation is also a bit mixed, sometimes it's called `Requestor` but usually `Requestors` :S

File: inspirehep/utils/tickets.py [hunk 8]
```diff
@@ -24,26 +24,222 @@
...
+                ... (according to RT fields)
+
+                Custom fields CF.{<CustomFieldName>} could be set
+                with keywords CF_CustomFieldName.
+    :returns: ID of the new ticket or ``-1``, if it fails
+    :rtype: integer
+    """
+    body = render_template(template_path, **template_context).strip()
+    subject = template_context.get("subject")
+    # FIXME: raise exception if subject is None ???
```
Comment (harunurhan) [inline inspirehep/utils/tickets.py]: it is gonna have `"None"` subject. Maybe we can also put `"no subject"` instead of raising.

File: inspirehep/utils/tickets.py [hunk 9]
```diff
@@ -24,26 +24,222 @@
...
+    """Resolves the given ticket
+
+    :type ticket_id: integer
+    """
+    try:
+        rt_instance.edit_ticket(
+            ticket_id=ticket_id,
+            Status="resolved",
+        )
+    except IndexError:
```
Comment (harunurhan) [inline inspirehep/utils/tickets.py]: It's actually not at all, it is a categorized error by `rt`

File: inspirehep/utils/ext.py (hunk 3 above)
Comment (jmartinm) [inline inspirehep/utils/ext.py]: I think its ok to pin the version from github. That way when we want to test locally the connection to the RT test instance, we will have this option available.

File: inspirehep/utils/tickets.py (hunk 8 above)
Comment (jmartinm) [inline inspirehep/utils/tickets.py]: Yes, ok to put a default value like `No Subject`

File: inspirehep/utils/tickets.py (hunk 9 above)
Comment (jmartinm) [inline inspirehep/utils/tickets.py]: What do you mean by `it is a categorized error by rt` ?

Comment (harunurhan) [inline inspirehep/utils/tickets.py]: sorry I meant it's NOT.

File: inspirehep/utils/ext.py (hunk 4 above)
Comment (harunurhan) [inline inspirehep/utils/ext.py]: no information found about expiration in 
https://rt-wiki.bestpractical.com/wiki/REST#Authentication

File: inspirehep/utils/tickets.py (hunk 7 above)
Comment (jmartinm) [inline inspirehep/utils/tickets.py]: So can you add in the docstring what is the expected format if you want to pass more than one requestor?

File: inspirehep/utils/tickets.py [hunk 10]
```diff
@@ -24,26 +24,234 @@
...
+                                **kwargs):
+    """Creates new RT ticket with a body that is rendered template
+
+    :param queue: where the ticket will be created
+    :type queue: string
+
+    :param requestors: username to set to requestors field of the ticket
+    :type requestors: string
+
+    :param template_path: path to the template for the ticket body
```
Comment (jmartinm) [inline inspirehep/utils/tickets.py]: :type string

File: inspirehep/utils/tickets.py [hunk 11]
```diff
@@ -24,26 +24,234 @@
...
+
+    :param template_context: context object to be used to render template
```
Comment (jmartinm) [inline inspirehep/utils/tickets.py]: :type dict

File: inspirehep/utils/tickets.py [hunk 12]
```diff
@@ -24,26 +24,234 @@
...
+                FinalPriority, TimeEstimated, Starts, Due,
+                ... (according to RT fields)
+
+                Custom fields CF.{<CustomFieldName>} could be set
+                with keywords CF_CustomFieldName.
+    :returns: ID of the new ticket or ``-1``, if it fails
+    :rtype: integer
+    """
+    body = render_template(template_path, **template_context).strip()
+    subject = template_context.get("subject")
```
Comment (jmartinm) [inline inspirehep/utils/tickets.py]: This is a bit weird, having to pass the Subject of the ticket in the template context. It would be better as a separate parameter.

Also you might need to encode in utf-8?

File: inspirehep/utils/tickets.py [hunk 13]
```diff
@@ -24,26 +24,234 @@
...
+
+    Takes a string like ``'17: CoolUser'`` and
+    returns ``{'id': '17', 'name': 'CoolUser'}``
+    """
+    parts = raw_item.split(": ")
+    return dict(id=parts[0], name=parts[1])
+
+
+def _strip_lines(multiline_string):
+    return "\n ".join(
```
Comment (jmartinm) [inline inspirehep/utils/tickets.py]: Docstring

File: inspirehep/modules/workflows/tasks/submission.py [hunk 14]
```diff
@@ -149,42 +149,19 @@ def _reply_ticket(obj, eng):
...
-        if keep_new:
-            # We keep the state as new
-            rt.edit_ticket(
-                ticket_id=ticket_id,
-                Status="new"
-            )
+            body = obj.extra_data.get("reason", "")
+            if body:
+                body = body.strip()
+                tickets.reply_ticket(ticket_id, body, keep_new)
```
Comment (jmartinm) [inline inspirehep/modules/workflows/tasks/submission.py]: I think you are already striping inside?

File: inspirehep/utils/tickets.py (hunk 10 above)
Comment (jacquerie) [inline inspirehep/utils/tickets.py]: ...although, given, that is a new docstring, what about using Google's style?

File: inspirehep/utils/tickets.py (hunk 12 above)
Comment (harunurhan) [inline inspirehep/utils/tickets.py]: utf-8 encoding done in the fuction that's called after.

If you think it would be better, I could make it as a separate param, I didn't mind because it's already in the context.

File: inspirehep/utils/tickets.py (hunk 13 above)
Comment (harunurhan) [inline inspirehep/utils/tickets.py]: Actually that's was workaround before, I am not sure what the put in the doc string that someone can't understand from the code.

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 15]
```diff
@@ -47,417 +45,37 @@
...
-    subschema = schema['properties']['titles']
-
+    mock_rt_instance = True
     data = {
         'titles': [
             {'title': 'Partial Symmetries of Weak Interactions'},
         ],
     }
     extra_data = {'ticket_id': 1}
-    assert validate(data['titles'], subschema) is None
```
Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: The calls to `validate` make sure that if the schema changes this test fails.

File: inspirehep/utils/tickets.py [hunk 16]
```diff
@@ -24,26 +24,242 @@
...
+def _get_all_of(query_type):
+    """Utility function to share the code for performing custom get all requests
+     and parsing the result
+
+    :param query_type: the type of quer, either ``'queue'`` or ``'user'``
+
+    :rtype: dict - with ``name (string)``, ``id (integer)`` properties
+    """
+    query_suffix = query_type + "?query="
+    url = join(rt_instance.url, "search", query_suffix)
```
Comment (jacquerie) [inline inspirehep/utils/tickets.py]: Don't build URLs with `os.path.join`: https://github.com/inspirehep/inspire-next/issues/2314.

File: inspirehep/utils/tickets.py [hunk 17]
```diff
@@ -24,26 +24,242 @@
...
+    return map(_query_result_item_id_name_mapper, lines)
+
+
+def _query_result_item_id_name_mapper(raw_item):
+    """Mapper function
+
+    Takes a string like ``'17: CoolUser'`` and
+    returns ``{'id': '17', 'name': 'CoolUser'}``
+    """
+    parts = raw_item.split(": ")
```
Comment (jacquerie) [inline inspirehep/utils/tickets.py]:
```python
id_, name = raw_item.split(': ')
return {'id': id_, 'name': name}
```

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 18]
```diff
@@ -47,417 +45,37 @@
...
-
-    assert expected == result['text']
+from mock import patch
 
-    expected = 'new'
-    result = mock_rt.get_ticket(1)
+from mocks import AttrDict, MockEng, MockFiles, MockObj, MockUser
 
-    assert expected == result['Status']
+import pytest
```
Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: Why are you moving around the imports that are not related to what you are doing? You also put them in an order which is inconsistent with the rest of the codebase...

File: tests/unit/utils/test_utils_tickets.py [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+"""Unit test for Tickets utils."""
```
Comment (jacquerie) [inline tests/unit/utils/test_utils_tickets.py]: We don't use docstrings in tests.

File: tests/unit/workflows/test_workflows_tasks_submission.py [hunk 20]
```diff
@@ -47,417 +45,37 @@
...
 
 @patch('inspirehep.modules.workflows.tasks.submission.User')
-@patch('inspirehep.modules.workflows.tasks.submission.get_instance')
-def test_reply_ticket_thanks_user_when_their_literature_submission_is_accepted(mock_get_instance, mock_user):
-    mock_rt = MockRT()
-    mock_rt.create_ticket(Queue='Test', Status='new')
-    mock_get_instance.return_value = mock_rt
+@patch('inspirehep.modules.workflows.tasks.submission.tickets.reply_ticket_with_template')
+@patch('inspirehep.modules.workflows.tasks.submission.rt_instance')
+def test_reply_ticket_calls_tickets_reply_with_template_when_template_is_set(mock_rt_instance, mock_reply_ticket_with_template, mock_user):
```
Comment (jmartinm) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: Test looks ok to me.

File: inspirehep/utils/ext.py [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+        """Initialize the application."""
+        self.rt_instance = self.create_rt_instance(app)
+        app.extensions["inspire-utils"] = self
+
+    def create_rt_instance(self, app):
+        """Make a RT instance and return it."""
+        url = app.config.get("CFG_BIBCATALOG_SYSTEM_RT_URL", "")
+        login = app.config.get("CFG_BIBCATALOG_SYSTEM_RT_DEFAULT_USER", "")
+        password = app.config.get(
+            "CFG_BIBCATALOG_SYSTEM_RT_DEFAULT_PWD", "")
```
Comment (jacquerie) [inline inspirehep/utils/ext.py]: Not to be fixed right now, but we should avoid Invenio 1 style configuration options. They are confusing (who is going to know what `BibCatalog` is one year from now?) and they do not conform to the usual patterns of Flask applications configuration options.

File: tests/unit/workflows/test_workflows_tasks_submission.py (hunk 15 above)
Comment (harunurhan) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: I actually didn't get why this was here, so in this I should add it to all tests?

Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_submission.py]: Actually you were right in removing this, because your implementation does no longer depend on the title being there. BTW, I'm going to merge this guy this weekend, so that you can continue working on this from Monday.

File: inspirehep/modules/workflows/tasks/submission.py [hunk 22]
```diff
@@ -100,28 +92,27 @@ def _create_ticket(obj, eng):
...
                     queue=queue,
                     subject=context.get('subject'),
                     requestors=user.email,
-                    body=body
                 )
             )
             return
 
+        recid = obj.extra_data.get("recid") or obj.data.get("control_number") \
+            or obj.data.get("recid")
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/submission.py]: You can remove `obj.data.get("recid")`, it's never there (`obj.data` complies with the schema).

**Comment by harunurhan** [comment]
@jmartinm  ping

**Comment by jmartinm** [comment]
> @jmartinm ping

LGTM. Can you rebase and we will merge this

**Comment by harunurhan** [comment]
@jmartinm there wasn't any conflicts but rebased again
